### ISSUE TITLE
Parse error when using keyword as identifier in DB2

### ISSUE THREAD

**Issue by nene** [sqlfluff/sqlfluff#5397]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When parsing the admittedly contrived (but valid) DB2 SQL:

```sql
SELECT having AS from FROM where;
```

Quoting [from DB2 documentation](https://www.ibm.com/docs/en/db2/11.5?topic=sql-reserved-schema-names-reserved-words):

> Keywords can be used as ordinary identifiers, except in a context where they could also be interpreted as SQL keywords. In such cases, the word must be specified as a delimited identifier. For example, COUNT cannot be used as a column name in a SELECT statement, unless it is delimited.

One can easily verify the validity of this code using the [dbfiddle](https://dbfiddle.uk/) tool.

### Expected Behaviour

Parses without failure.

### Observed Behaviour

Syntax error occurs:

```sh
$sqlfluff parse test.sql --dialect db2

[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'SELECT'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    column_reference:
[L:  1, P:  8]      |                        naked_identifier:                     'having'
[L:  1, P: 14]      |                [META] dedent:
[L:  1, P: 14]      |                whitespace:                                   ' '
[L:  1, P: 15]      |                unparsable:                                   !! Expected: 'Nothing here.'
[L:  1, P: 15]      |                    word:                                     'AS'
[L:  1, P: 17]      |            whitespace:                                       ' '
[L:  1, P: 18]      |            from_clause:
[L:  1, P: 18]      |                keyword:                                      'from'
[L:  1, P: 22]      |                whitespace:                                   ' '
[L:  1, P: 23]      |                from_expression:
[L:  1, P: 23]      |                    [META] indent:
[L:  1, P: 23]      |                    from_expression_element:
[L:  1, P: 23]      |                        table_expression:
[L:  1, P: 23]      |                            table_reference:
[L:  1, P: 23]      |                                naked_identifier:             'FROM'
[L:  1, P: 27]      |                    [META] dedent:
[L:  1, P: 27]      |            whitespace:                                       ' '
[L:  1, P: 28]      |            unparsable:                                       !! Expected: 'Nothing here.'
[L:  1, P: 28]      |                word:                                         'where'
[L:  1, P: 33]      |    statement_terminator:                                     ';'
[L:  1, P: 34]      |    newline:                                                  '\n'
[L:  2, P:  1]      |    [META] end_of_file:

==== parsing violations ====
L:   1 | P:  15 |  PRS | Line 1, Position 15: Found unparsable section: 'AS'
L:   1 | P:  28 |  PRS | Line 1, Position 28: Found unparsable section: 'where'
```

### How to reproduce

Just parse or lint the code example above.

### Dialect

DB2

### Version

- SQLFluff 2.3.5
- Python 3.11.5

### Configuration

No, I don't have any configuration files.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by nene** [comment]
Some additional thoughts:

Admittedly, not really having any fully reserved keywords makes DB2 quite distinct among SQL dialects. Even PostgreSQL, which is quite permissive doesn't allow the above query. I recall hearing about some other SQL dialect also having this property, but I don't recall what it was.

Also, one probably deserves to be smacked with parser error when writing SQL that uses such common keywords as identifiers. :)

**Comment by peterbud** [comment]
Hi @nene ,

Thank you for the detailed report and for the "admittedly contrived" (but very interesting!) example. You’re absolutely right that DB2’s permissiveness with keywords makes it a bit of a unicorn among SQL dialects.

While we strive for high compatibility, supporting this specific behavior presents a significant technical challenge for our current architecture:

* **Grammar Constraints**: Our parser relies on a set of reserved keywords to maintain a deterministic and performant grammar. Allowing keywords like FROM or WHERE to act as unquoted identifiers would introduce widespread ambiguities that our current engine isn't designed to resolve without a massive increase in lookahead complexity.

* **Maintenance Trade-off**: As you noted, even PostgreSQL draws the line here. We’ve found that enforcing standard reserved keywords allows us to keep the codebase maintainable and less prone to regressions in more common use cases.

**Recommendation:**
For the time being, we require these keywords to be provided as delimited identifiers (e.g., "FROM") to be parsed correctly.
